### PR TITLE
gh-95494: Fix transport EOF handling in OpenSSL 3.0

### PR DESCRIPTION
**Pull request by davidben**
GH-25309 enabled SSL_OP_IGNORE_UNEXPECTED_EOF by default, with a comment
that it restores OpenSSL 1.1.1 behavior, but this wasn't quite right.
That option causes OpenSSL to treat transport EOF as the same as
close_notify (i.e. SSL_ERROR_ZERO_RETURN), whereas Python actually has
distinct SSLEOFError and SSLZeroReturnError exceptions. (The latter is
usually mapped to a zero return from read.) In OpenSSL 1.1.1, the ssl
module would raise them for transport EOF and close_notify,
respectively. In OpenSSL 3.0, both act like close_notify.

Fix this by, instead, just detecting SSL_R_UNEXPECTED_EOF_WHILE_READING
and mapping that to the other exception type.

There doesn't seem to have been any unit test of this error, so fill in
the missing one. This had to be done with the BIO path because it's
actually slightly tricky to simulate a transport EOF with Python's fd
based APIs. (If you instruct the server to close the socket, it gets
confused, probably because the server's SSL object is still referencing
the now dead fd?)

* Issue: gh-95494

### CODE REVIEW

**Comment by davidben** [comment]
EOF handling in OpenSSL has gotten unnecessarily hairy, largely due to upstream getting the design of `BIO_read_ex` and `SSL_read_ex` wrong. I ran into this while trying to work through this mess. We can probably simplify the EOF logic elsewhere in that function, but I'll leave that to when I've disentangled it more. This fix seems to be separable.

**Comment by davidben** [comment]
Seem to have broken a few tests. Odd. Will take a look and update this.

**Comment by davidben** [comment]
> Seem to have broken a few tests. Odd. Will take a look and update this.

Fixed, hopefully. The issue was I needed the new exception to have `PY_SSL_ERROR_EOF`.

**Review by tiran — CHANGES_REQUESTED** [review]
Fantastic! Thank you David.

Could you please extend the blurb and mention that :attr:`~ssl.SSLContext.options` no longer has `OP_IGNORE_UNEXPECTED_EOF`?

**Comment by davidben** [comment]
I have made the requested changes; please review again

Let me know if I did that right. I went and added the various role markers to the other links too.

**Comment by davidben** [comment]
Anything left to do on this CL on my end?

**Comment by davidben** [comment]
@tiran Anything left for me to do before this is mergable?

**Review by Yhg1s — APPROVED** [review]

**Comment by Yhg1s** [comment]
Merging, since Christian is taking a break and the only open comment is about the news entry, which was addressed.

**Comment by Yhg1s** [comment]
The only question now is whether to backport this to 3.11 and 3.10. I think it makes sense to backport, but @pablogsal, @ambv, what do you think? (This changes the exception raised when using OpenSSL 3.0 and a TLS connection is unexpectedly terminated, but changes it to behave more like OpenSSL 1.1.)

**Comment by ambv** [comment]
This is entirely in Pablo's hands. He's right that with 3.10 being this far down the line, it's disruptive now.

However! 3.11 has only seen two bugfix releases so far. Maybe that's not too late to course-correct? People are still not using 3.11 very widely in the wild. If Black is any indication, pip installs from 3.11 account for barely 8% of the total (3.7 - 3.11). 3.10 is still 3X more popular than 3.11, and so is 3.9. In fact, 3.8 beats 3.11 by five times.

Numpy presents an even colder image of 3.11 adoption. Downloads from 3.11 constitute barely 2.0% - 2.3% of the total. 3.11 barely beats 2.7 there. 3.10 beats it by almost 4X, 3.9 by over 8X, 3.8 by a whopping 14.4X, and 3.7 by close to 16X.

Anaconda also doesn't use 3.11 yet, [it's available in the main repository](https://www.anaconda.com/blog/upcoming-releases-anaconda-distribution-2023-03-and-beyond) but they only just released 3.10 support so nobody is using that in the wild unless they opt-in.

What I'm saying is: In my opinion **3.11 can still fix things with relatively little consequence to external users**.

**Comment by ambv** [comment]
Some bug fixes necessarily break what people used before. In my time I recall doing a small number of breaking fixes when the given Python version was the newest maintenance release. The way I thought about this at the time was this (version numbers adapted to the current situation):

- 3.12 is still being developed
- 3.11.3 will change behavior
- therefore, a library maintainer only has to say `if sys.version_info[:3] >= (3, 11, 3):` to guard the change -- this covers 3.12 as well

Changing older maintenance releases would create annoying version cherry-picking for maintainers. This one in the example above? Not at all different from `if sys.version_info[:2] >= (3, 12)`. It's still just one check.

**Comment by Yhg1s** [comment]
Note that while this introduces a change _when using OpenSSL 3.0_, it actually makes it behave more like OpenSSL 1.1. There's no change when using OpenSSL 1.1. The only way (I can think of) this would affect user code is when they have only been using OpenSSL 3.0, and they're are handling SSLZeroReturnError but not EOFError. The code would already be problematic if run with OpenSSL 1.1 instead. I feel like the adoption of the Python version hardly matters in that equation.

**Comment by davidben** [comment]
(To clarify, I have no particular stake in this being classified as a security issue or not. In so far as it is one, I doubt many, if any, applications care. Just realized it might matter to some folks, so thought I'd mention in case you all care.)

**Comment by pablogsal** [comment]
After reflecting a bit more about this I think I agree with @ambv and I would be ok backporting it to 3.11

**Pull request merged**
